File: core/src/main/java/io/github/msameer0/rhythmicrush/game/gameplay/blocks/Block.java
package io.github.msameer0.rhythmicrush.game.gameplay.blocks;

import com.badlogic.gdx.math.Rectangle;

import io.github.msameer0.rhythmicrush.game.GameWorld;
import io.github.msameer0.rhythmicrush.game.gameplay.players.AbstractPlayer;

/**
 * Represents a physical platform or obstacle within the game world.
 * <p>
 * The {@code Block} class manages its own position, dimensions, and collision bounds.
 * it provides logic for horizontal movement (scrolling) and complex collision
 * detection with {@link AbstractPlayer} instances, including landing on top,
 * bumping into the bottom, and triggering death upon side impacts.
 * </p>
 * <p>
 * This class is designed to be compatible with object pooling to optimize memory
 * usage during gameplay.
 * </p>
 */
public class Block {

    protected float x, y;
    protected float width, height;
    protected Rectangle bounds;
    protected BlockType type;

    /**
     * Constructs a new empty Block.
     * <p>
     * This no-arg constructor is primarily intended for object pooling.
     * The {@link #init(float, float, float, BlockType)} method must be called
     * before the block can be used in the game world.
     * </p>
     */
    public Block() {
        bounds = new Rectangle();
    }

    /**
     * Constructs a new Block with the specified position and size using the default block type.
     *
     * @param x    the x-coordinate of the block
     * @param y    the y-coordinate of the block
     * @param size the width and height of the block
     */
    public Block(float x, float y, float size) {
        this(x, y, size, BlockType.DEFAULT);
    }

    /**
     * Constructs a new Block with the specified position, size, and type.
     *
     * @param x    the x-coordinate of the block
     * @param y    the y-coordinate of the block
     * @param size the width and height of the block
     * @param type the type of the block defining its appearance or behavior
     */
    public Block(float x, float y, float size, BlockType type) {
        this.x = x;
        this.y = y;
        this.width = size;
        this.height = size;
        this.type = type;
        bounds = new Rectangle(x, y, width, height);
    }

    /**
     * Reinitialise this block for reuse from the pool.
     */
    public Block init(float x, float y, float size, BlockType type) {
        this.x = x;
        this.y = y;
        this.width = size;
        this.height = size;
        this.type = type;
        bounds.set(x, y, width, height);
        return this;
    }

    public void updatePosition(float scrollSpeed, float delta) {
        x -= scrollSpeed * delta;
        updateBounds();
    }

    public Rectangle getBounds() {
        return bounds;
    }

    public void updateBounds() {
        bounds.setPosition(x, y);
    }

    /**
     * Handles collision detection and response between the player and this block.
     * <p>
     * This method calculates the overlap between the player's bounding box and the block's
     * bounds to determine the direction of impact. It supports the following behaviors:
     * <ul>
     *     <li><b>Landing:</b> If the player hits the top of the block while falling, they are
     *     snapped to the surface and marked as grounded.</li>
     */
    public void tryTouch(AbstractPlayer player) {
        Rectangle playerRect = player.getBounds();
        if (!playerRect.overlaps(bounds)) return;

        float playerBottom = playerRect.y;
        float playerTop = playerRect.y + playerRect.height;
        float playerLeft = playerRect.x;
        float playerRight = playerRect.x + playerRect.width;

        float blockTop = bounds.y + bounds.height;
        float blockBottom = bounds.y;
        float blockLeft = bounds.x;
        float blockRight = bounds.x + bounds.width;

        float overlapTop = blockTop - playerBottom;
        float overlapBottom = playerTop - blockBottom;
        float overlapLeft = playerRight - blockLeft;
        float overlapRight = blockRight - playerLeft;

        float minOverlap = overlapTop;
        if (overlapBottom < minOverlap) minOverlap = overlapBottom;
        if (overlapLeft < minOverlap) minOverlap = overlapLeft;
        if (overlapRight < minOverlap) minOverlap = overlapRight;

        if (minOverlap == overlapTop && player.velocityY <= 0) {
            player.setY(blockTop);
            player.setVelocityY(0);
            player.setGrounded(true);
            return;
        }

        if (minOverlap == overlapBottom && player.velocityY >= 0 && player.isSafeFromBelow()) {
            player.setY(blockBottom - player.height);
            player.setVelocityY(0);
            return;
        }

        float hMargin = playerRect.width * 0.25f;
        float vMargin = playerRect.height * 0.25f;

        if (playerRight - hMargin > blockLeft && playerLeft + hMargin < blockRight &&
            playerTop - vMargin > blockBottom && playerBottom + vMargin < blockTop) {
            GameWorld world = player.getWorld();
            if (world != null) world.playerDied();
        }
    }

    /**
     * Gets the x-coordinate of the block.
     *
     * @return the x-coordinate
     */
    public float getX() {
        return x;
    }

    /**
     * Returns the y-coordinate of the block.
     *
     * @return the y-coordinate of the block
     */
    public float getY() {
        return y;
    }

    /**
     * Returns the width of the block.
     *
     * @return the width of the block
     */
    public float getWidth() {
        return width;
    }

    /**
     * @return the height of the block
     */
    public float getHeight() {
        return height;
    }

    /**
     * Gets the type of this block, which defines its appearance or behavior.
     *
     * @return the {@link BlockType} of the block
     */
    public BlockType getType() {
        return type;
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/gameplay/blocks/Block.kt
package io.github.msameer0.rhythmicrush.game.gameplay.blocks

import com.badlogic.gdx.math.Intersector
import com.badlogic.gdx.math.Polygon
import com.badlogic.gdx.math.Rectangle
import io.github.msameer0.rhythmicrush.game.engine.Rotatable
import io.github.msameer0.rhythmicrush.game.engine.Tickable
import io.github.msameer0.rhythmicrush.game.gameplay.players.AbstractPlayer
import io.github.msameer0.rhythmicrush.game.registries.Registry

@Registry(id = "block")
/**
 * Base class for all physical blocks in the game that the player can collide with or walk on.
 */
open class Block : Tickable, Rotatable {
    override var rotation: Float = 0f
    var x: Float = 0f
    var worldX: Float = 0f
    var y: Float = 0f
    var width: Float = 0f
    var height: Float = 0f
    var bounds: Rectangle = Rectangle()
    var type: BlockType
    var untouchable: Boolean = false

    constructor() {
        this.type = BlockType.DEFAULT
    }

    constructor(x: Float, y: Float, size: Float) : this(x, y, size, BlockType.DEFAULT)

    constructor(x: Float, y: Float, size: Float, type: BlockType) {
        this.x = x
        this.y = y
        this.width = size
        this.height = size
        this.type = type
        bounds = Rectangle(x, y, width, height)
    }

    open fun init(x: Float, y: Float, size: Float, type: BlockType, rotation: Float): Block {
        this.x = x
        this.worldX = x
        this.y = y
        this.width = size
        this.height = size
        this.type = type
        this.rotation = rotation
        this.untouchable = false
        bounds.set(x, y, width, height)
        return this
    }

    open fun init(x: Float, y: Float, size: Float, type: BlockType): Block {
        return init(x, y, size, type, 0f)
    }

    fun updatePosition(worldScrolled: Float) {
        x = worldX - worldScrolled
        updateBounds()
    }

    fun updateBounds() {
        bounds.setPosition(x, y)
    }

    private val blockPolygon = Polygon()
    private var polygonWidth = -1f
    private var polygonHeight = -1f

    open fun reset() {
        this.x = 0f
        this.worldX = 0f
        this.y = 0f
        this.width = 0f
        this.height = 0f
        this.type = BlockType.DEFAULT
        this.rotation = 0f
        this.untouchable = false
        this.bounds.set(0f, 0f, 0f, 0f)
    }

    open fun tryTouch(player: AbstractPlayer) {
        if (untouchable) return

        val playerRect = player.getBounds()
        if (!playerRect.overlaps(bounds)) return

        val playerBottom = playerRect.y
        val playerTop = playerRect.y + playerRect.height
        val playerLeft = playerRect.x
        val playerRight = playerRect.x + playerRect.width

        val blockTop = bounds.y + bounds.height
        val blockBottom = bounds.y
        val blockLeft = bounds.x
        val blockRight = bounds.x + bounds.width

        val overlapTop = blockTop - playerBottom
        val overlapBottom = playerTop - blockBottom
        val overlapLeft = playerRight - blockLeft
        val overlapRight = blockRight - playerLeft

        var minOverlap = overlapTop
        if (overlapBottom < minOverlap) minOverlap = overlapBottom
        if (overlapLeft < minOverlap) minOverlap = overlapLeft
        if (overlapRight < minOverlap) minOverlap = overlapRight

        val flipped = player.isGravityFlipped()

        if (!flipped) {
            if (minOverlap == overlapTop && player.velocityY <= 0) {
                player.setY(blockTop)
                player.setVelocityY(0f)
                player.setGrounded(true)
                return
            }

            if (minOverlap == overlapBottom && player.velocityY >= 0 && player.isSafeFromBelow()) {
                player.setY(blockBottom - player.height)
                player.setVelocityY(0f)
                return
            }
        } else {
            if (minOverlap == overlapBottom && player.velocityY >= 0) {
                player.setY(blockBottom - player.height)
                player.setVelocityY(0f)
                player.setGrounded(true)
                return
            }

            if (minOverlap == overlapTop && player.velocityY <= 0 && player.isSafeFromBelow()) {
                player.setY(blockTop)
                player.setVelocityY(0f)
                return
            }
        }

        if (polygonWidth != width || polygonHeight != height) {
            var verts = blockPolygon.vertices
            if (verts == null || verts.size != 8) {
                verts = FloatArray(8)
                blockPolygon.vertices = verts
            }
            verts[0] = 0f; verts[1] = 0f
            verts[2] = width; verts[3] = 0f
            verts[4] = width; verts[5] = height
            verts[6] = 0f; verts[7] = height
            polygonWidth = width
            polygonHeight = height
        }
        blockPolygon.setPosition(x, y)
        blockPolygon.setOrigin(width / 2f, height / 2f)
        blockPolygon.rotation = rotation

        if (Intersector.overlapConvexPolygons(player.getInnerPolygon(), blockPolygon)) {
            val world = player.getWorld()
            world?.playerDied()
        }
    }

    override fun tick(delta: Float) {
    }
}
